File: android/app/src/main/java/app/notomorrow/feature/workout/WorkoutStrings.kt
package app.notomorrow.feature.workout

import androidx.compose.runtime.Composable
import androidx.compose.ui.res.stringResource
import app.notomorrow.data.entity.ExerciseEntity
import app.notomorrow.designsystem.appLocale
import app.notomorrow.util.Localizer
import app.notomorrow.util.NtKeys
import app.notomorrow.util.S
import java.util.Locale

/**
 * Copy helpers for the Train tab — 1:1 port of
 * `NoTomorrow/Features/Workout/WorkoutStrings.swift`.
 *
 * Two halves survive the port:
 *
 *  * the **counted nouns** (`WorkoutStrings.counted`), whose Polish CLDR selector already
 *    lives in [NtKeys.pluralCategory] because the catalog carries `one` / `few` / `many` as
 *    three plain strings rather than as a `<plurals>`;
 *  * the **muscle / equipment vocabulary**, whose runtime-built keys
 *    (`"muscleName.\(slug)"`) are the exhaustive `when` maps in [NtKeys].
 *
 * `WorkoutStrings.fold` is **not** here: it is `ExerciseLibrary.fold`, next to the
 * `sorted`/`filter` pair that uses it.
 *
 * Every function comes in two flavours: a pure one taking a [Localizer] (view models,
 * services, tests) and a `@Composable` one reading `stringResource` (views).
 */
object WorkoutStrings {

    // MARK: - Plurals

    /** "5 exercises" / "5 ćwiczeń" */
    fun exercises(n: Int, strings: Localizer): String = strings.string(NtKeys.exerciseCount(n), n)

    /** "3 sets" / "3 serie" */
    fun sets(n: Int, strings: Localizer): String = strings.string(NtKeys.setCount(n), n)

    /** "2 PRs" — the dashboard key; Polish keeps "PR" invariant. */
    fun prs(n: Int, strings: Localizer): String = strings.string(S.dashboard_prs, n)

    /** "7 results" */
    fun results(n: Int, strings: Localizer): String = strings.string(S.exercises_results, n)

    /** "Add 3" */
    fun add(n: Int, strings: Localizer): String = strings.string(S.exercises_addCount, n)

    /** Create "&lt;query&gt;" as a new exercise */
    fun create(query: String, strings: Localizer): String = strings.string(S.exercises_create, query)

    // MARK: - Muscles & equipment

    /**
     * Localized muscle name ("chest" → "Chest" / "Klatka"). An unknown value falls back to
     * the capitalized raw string, exactly as the iOS `value == key` check does.
     */
    fun muscle(raw: String, strings: Localizer, locale: Locale): String =
        NtKeys.muscle(raw)?.let(strings::string) ?: capitalized(raw, locale)

    fun equipment(raw: String, strings: Localizer, locale: Locale): String =
        NtKeys.equipment(raw)?.let(strings::string) ?: capitalized(raw, locale)

    /**
     * "Chest · Triceps, Shoulders" — primary muscles, then the first two secondaries;
     * the equipment stands in when there are no secondaries.
     */
    fun subtitle(exercise: ExerciseEntity, strings: Localizer, locale: Locale): String {
        val primary = exercise.primaryMuscles.joinToString(", ") { muscle(it, strings, locale) }
        val secondary = exercise.secondaryMuscles.take(2).joinToString(", ") { muscle(it, strings, locale) }
        val parts = buildList {
            if (primary.isNotEmpty()) add(primary)
            if (secondary.isNotEmpty()) {
                add(secondary)
            } else {
                val equipment = exercise.equipment
                if (!equipment.isNullOrEmpty()) add(equipment(equipment, strings, locale))
            }
        }
        return parts.joinToString(" · ")
    }

    /** `String.capitalized(with: .current)` — first letter only, under the app locale. */
    private fun capitalized(raw: String, locale: Locale): String =
        raw.replaceFirstChar { if (it.isLowerCase()) it.titlecase(locale) else it.toString() }
}

// MARK: - Composable flavours

/** [WorkoutStrings.exercises] inside composition. */
@Composable
fun workoutExerciseCount(n: Int): String = stringResource(NtKeys.exerciseCount(n), n)

/** [WorkoutStrings.sets] inside composition. */
@Composable
fun workoutSetCount(n: Int): String = stringResource(NtKeys.setCount(n), n)

/** [WorkoutStrings.muscle] inside composition. */
@Composable
fun workoutMuscleName(raw: String): String {
    val locale = appLocale()
    val id = NtKeys.muscle(raw)
    return if (id != null) {
        stringResource(id)
    } else {
        raw.replaceFirstChar { if (it.isLowerCase()) it.titlecase(locale) else it.toString() }
    }
}

File: NoTomorrow/Features/Workout/WorkoutStrings.swift
import Foundation

/// Copy helpers for the Train tab: counted nouns with Polish plural categories (one / few / many)
/// and the muscle / equipment vocabulary from free-exercise-db mapped onto catalog keys.
enum WorkoutStrings {

    // MARK: Plurals

    /// CLDR plural category as used by Polish and English (the English catalog values for `few` and `many` are identical).
    private static func pluralSuffix(_ n: Int) -> String {
        if n == 1 { return "one" }
        let mod10 = n % 10, mod100 = n % 100
        if (2...4).contains(mod10) && !(12...14).contains(mod100) { return "few" }
        return "many"
    }

    private static func counted(_ base: String, _ n: Int) -> String {
        let key = "\(base).\(pluralSuffix(n))"
        let pattern = String(localized: String.LocalizationValue(key))
        return String(format: pattern, locale: .current, n)
    }

    /// "5 exercises" / "5 ćwiczeń"
    static func exercises(_ n: Int) -> String { counted("workout.exerciseCount", n) }

    /// "3 sets" / "3 serie"
    static func sets(_ n: Int) -> String { counted("workout.setCount", n) }

    /// "2 PRs" — existing dashboard key, Polish keeps "PR" invariant.
    static func prs(_ n: Int) -> String {
        String(format: String(localized: "dashboard.prs"), locale: .current, n)
    }

    /// "7 results"
    static func results(_ n: Int) -> String {
        String(format: String(localized: "exercises.results"), locale: .current, n)
    }

    /// "Add 3"
    static func add(_ n: Int) -> String {
        String(format: String(localized: "exercises.addCount"), locale: .current, n)
    }

    /// Create "<query>" as a new exercise
    static func create(_ query: String) -> String {
        String(format: String(localized: "exercises.create"), locale: .current, query)
    }

    // MARK: Muscles & equipment

    private static func slug(_ raw: String) -> String {
        raw.lowercased()
            .replacingOccurrences(of: " ", with: "_")
            .replacingOccurrences(of: "-", with: "_")
    }

    /// Localized muscle name ("chest" → "Chest" / "Klatka"). Unknown values fall back to a capitalized raw string.
    static func muscle(_ raw: String) -> String {
        let key = "muscleName.\(slug(raw))"
        let value = String(localized: String.LocalizationValue(key))
        return value == key ? raw.capitalized(with: .current) : value
    }

    static func equipment(_ raw: String) -> String {
        let key = "equipment.\(slug(raw))"
        let value = String(localized: String.LocalizationValue(key))
        return value == key ? raw.capitalized(with: .current) : value
    }

    /// "Chest · Triceps, Shoulders" — primary muscles, then secondaries; equipment when there are no secondaries.
    static func subtitle(for exercise: Exercise) -> String {
        let primary = exercise.primaryMuscles.map(muscle).joined(separator: ", ")
        let secondary = exercise.secondaryMuscles.prefix(2).map(muscle).joined(separator: ", ")
        var parts: [String] = []
        if !primary.isEmpty { parts.append(primary) }
        if !secondary.isEmpty {
            parts.append(secondary)
        } else if let equipment = exercise.equipment, !equipment.isEmpty {
            parts.append(self.equipment(equipment))
        }
        return parts.joined(separator: " · ")
    }

    /// Case- and diacritic-insensitive search key. Foundation strips the marks from ą ę ó ś ż ź ć ń but keeps ł
    /// (it has no decomposition), so ł is flattened by hand: "lawka" finds "Ławka". Locale-independent, like Android.
    static func fold(_ text: String) -> String {
        text.folding(options: [.caseInsensitive, .diacriticInsensitive, .widthInsensitive], locale: nil)
            .replacingOccurrences(of: "ł", with: "l")
            .replacingOccurrences(of: "Ł", with: "l")
    }
}
